File: dotNET/PFProSDK/src/PayPal/Payments/DataObjects/PaymentCard.cs
#region "Imports"

using System;
using PayPal.Payments.Common.Utility;
using PayPal.Payments.Common.Exceptions;

#endregion

namespace PayPal.Payments.DataObjects
{
	/// <summary>
	/// This abstract class serves as base
	/// class for Card Payment devices.
	/// </summary>
	/// <remarks>This class can be extended to create a new payment device type.</remarks>
	public class PaymentCard : PaymentDevice
	{
		#region "Member Variables"

		/// <summary>
		/// Card Expiry Date
		/// </summary>
		private String mExpDate;

		/// <summary>
		/// Card CVV2 code
		/// </summary>
		private String mCvv2;
		
		/// <summary>
		/// CardStart
		/// </summary>
		private String mCardStart;

		/// <summary>
		/// Card Issue
		/// </summary>
		private String mCardIssue;

        /// <summary>
        /// Card on File
        /// </summary>
        private String mCardonFile;

        /// <summary>
        /// Card on File Transaction ID
        /// </summary>
        private String mTxId;

		/// <summary>
		/// Payment Account Reference
		/// </summary>
		private String mParId;




		#endregion

		#region "Constructors"

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="Acct">Card number</param>
		/// <param name="ExpDate">Card expiry date</param>
		/// <remarks>Abstract class. Instance cannot be created directly.</remarks>
		public PaymentCard(String Acct, String ExpDate) : base(Acct)
		{
			mExpDate = ExpDate;
		}

		#endregion

		#region "Properties"

		/// <summary>
		/// Gets, Sets  Cvv2
		/// </summary>
		/// <remarks>
		/// Card validation code. This is the 3 or 4 digit code 
		/// present at the back side of the card.
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>CVV2</code>
		/// </remarks>
		public String Cvv2
		{
			get { return mCvv2; }
			set { mCvv2 = value; }
		}
		/// <summary>
		/// Gets, Sets  CardStart
		/// </summary>
		/// <remarks>
		/// Used for Switch/Solo Cards.
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>CARDSTART</code>
		/// </remarks>
		public String CardStart
		{
			get { return mCardStart; }
			set { mCardStart= value; }
		}
		/// <summary>
		/// Gets, Sets  CardIssue
		/// </summary>
		/// <remarks>
		/// Used for Switch/Solo Cards. 
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>CARDISSUE</code>
		/// </remarks>
		public String CardIssue
		{
			get { return mCardIssue; }
			set { mCardIssue = value; }
		}

        /// <summary>
        /// Gets, Sets  Card on File
        /// </summary>
        /// <remarks>
        /// Used to flag if the card is on file.
        /// <para>Maps to Payflow Parameter:</para>
        /// <code>CARDONFILE</code>
        /// </remarks>
        public String CardonFile
        {
            get { return mCardonFile; }
            set { mCardonFile = value; }
        }

        /// <summary>
        /// Gets, Sets  Transaction ID
        /// </summary>
        /// <remarks>
        /// Used to flag if the card is on file.
        /// <para>Maps to Payflow Parameter:</para>
        /// <code>TXID</code>
        /// </remarks>
        public String TxId
        {
            get { return mTxId; }
            set { mTxId = value; }
        }

		/// <summary>
		/// Gets, Sets PAR
		/// </summary>
		/// <remarks>
		/// Reference number assigned to each unique Primary Account Number (PAN).
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>PAR</code>
		/// </remarks>
		public String ParId
		{
			get { return mParId; }
			set { mParId = value; }
		}

		#endregion

		#region "Core functions"

		/// <summary>
		/// Generates the transaction request.
		/// </summary>
		internal override void GenerateRequest()
		{
			try
			{
				base.GenerateRequest();
				RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_EXPDATE, mExpDate));
				RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_CVV2, mCvv2));
				RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_CARDSTART, mCardStart));
				RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_CARDISSUE, mCardIssue));
                RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_CARDONFILE, mCardonFile));
                RequestBuffer.Append(PayflowUtility.AppendToRequest(PayflowConstants.PARAM_TXID, mTxId));
            }
			catch (BaseException)
			{
				throw;
			}
			catch (Exception Ex)
			{
				DataObjectException DEx = new DataObjectException(Ex);
				throw DEx;
			}
		}
		#endregion
	}
}

File: java/src/sdk/dataobjects/paypal/payflow/PaymentCard.java
package paypal.payflow;

/**
 * This abstract class serves as base class for Card Payment devices.
 * This class can be extended to create a new payment device type.
 */
abstract class PaymentCard extends PaymentDevice {
    /**
     * Card Expiry Date
     */
    private String expDate;
    /**
     * Card CVV2 code
     */
    private String cvv2;
    /**
     * Card Sart used for Swicth/Solo cards.
     */
    private String cardStart;
    /**
     * Card Issue used for Swicth/Solo cards.
     */
    private String cardIssue;
    /**
     * Flag card as stored credential (card on file).
     */
    private String cardOnFile;
    /**
     * Transaction Id for stored credential (card on file).
     */
    private String txId;
    /**
     * Payment Account Reference
     */
    private String par;


    /**
     * Constructor
     *
     * @param acct    Card number
     * @param expDate Card expiry date
     */
    protected PaymentCard(String acct, String expDate) {
        super(acct);
        this.expDate = expDate;
    }

    /**
     * Gets the CVV2 value.
     * Card validation code. This is the 3 or 4 digit code present at the back side of the card.
     *
     * @return CVV2 value
     *          * <p>
     *  Maps to Payflow Parameters as follows: CVV2
     * </p>
     */
    public String getCvv2() {
        return cvv2;
    }

    /**
     * Sets the CVV2 value.
     * Card validation code. This is the 3 or 4 digit code present at the back side of the card.
     *
     * @param cvv2 CVV2 value
     *              * <p>
     *  Maps to Payflow Parameters as follows: CVV2
     * </p>
     */
    public void setCvv2(String cvv2) {
        this.cvv2 = cvv2;
    }

    /**
     * Gets the cardIssue
     * <p> Used in Switch/Solo cards </p>
     *
     * @return cardIssue String
     *          * <p>
     *  Maps to Payflow Parameter : CARDISSUE
     * </p>
     */
    public String getCardIssue() {
        return cardIssue;
    }

    /**
     * Sets the cardIssue
     * <p> Used in Switch/Solo cards </p>
     *
     * @param cardIssue String
     *                   * <p>
     *  Maps to Payflow Parameter : CARDISSUE
     * </p>
     */
    public void setCardIssue(String cardIssue) {
        this.cardIssue = cardIssue;
    }

    /**
     * Gets the cardStart
     * <p> Used in Switch/Solo cards </p>
     *
     * @return cardStart String
     *          * <p>
     *  Maps to Payflow Parameter : CARDSTART
     * </p>
     */
    public String getCardStart() {
        return cardStart;
    }

    /**
     * Sets the cardStart
     * <p> Used in Switch/Solo cards </p>
     *
     * @param cardStart String
     *                   * <p>
     *  Maps to Payflow Parameter : CARDSTART
     * </p>
     */
    public void setCardStart(String cardStart) {
        this.cardStart = cardStart;
    }

    /**
     * Gets the cardOnFile
     * <p> Used to store credit card (stored credential)  </p>
     *
     * @return cardOnFile String
     *          * <p>
     *  Maps to Payflow Parameter : CARDONFILE
     * </p>
     */
    public String getCardOnFile() {
        return cardOnFile;
    }

    /**
     * Sets the cardOnFile
     * <p> Used to store credit card (stored credential) </p>
     *
     * @param cardOnFile String
     *                   * <p>
     *  Maps to Payflow Parameter : CARDONFILE
     * </p>
     */
    public void setCardOnFile(String cardOnFile) {
        this.cardOnFile= cardOnFile;
    }

    /**
     * Gets the TxId
     * <p> The transaction Id to reference a stored credential.</p>
     *
     * @return txId String
     *          * <p>
     *  Maps to Payflow Parameter : TXID
     * </p>
     */
    public String getTxId() {
        return txId;
    }
    /**
     * Sets the Transaction Id (stored credential).
     * <p> The transaction Id to reference a stored credential.</p>
     *
     * @param txId String
     *                   * <p>
     *  Maps to Payflow Parameter : TXID
     * </p>
     */
    public void setTxId(String txId) {
        this.txId = txId;
    }


    /**
     * Gets the par
     * <p>unique Primary Account Number (PAN)</p>
     *
     * @return par String
     *          * <p>
     *  Maps to Payflow Parameter : PAR
     * </p>
     */
    public String getParId() {
        return par;
    }
    /**
     * Sets the Payment Account Reference.
     * <p>A non-financial reference number assigned to each unique Primary Account Number (PAN) and mapped to all its affiliated Payment Tokens.</p>
     *
     * @param par String
     *                   * <p>
     *  Maps to Payflow Parameter : PAR
     * </p>
     */
    public void setPar(String par) {
        this.par = par;
    }

    /**
     * Generates the transaction request.
     */
    protected void generateRequest() {
        super.generateRequest();
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_EXPDATE, expDate));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_CVV2, cvv2));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_CARDSTART, cardStart));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_CARDISSUE, cardIssue));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_CARDONFILE, cardOnFile));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_TXID, txId));
        super.getRequestBuffer().append(PayflowUtility.appendToRequest(PayflowConstants.PARAM_PAR, par));
    }
}
